### ISSUE TITLE
Libreswan 3.15 - Failed to compile

### ISSUE THREAD

**Issue by hwdsl2** [libreswan/libreswan#36]
Operating System: CentOS 6.6
Kernel version: 2.6.32-504.30.3.el6.x86_64
LIbreswan version: 3.15

Compile error log:
```
root@server > make programs
... ...
In file included from /opt/src/libreswan-3.15/programs/pluto/demux.h:21,
                 from /opt/src/libreswan-3.15/programs/pluto/connections.c:60:
/opt/src/libreswan-3.15/programs/pluto/server.h:21:52: error: event2/event.h: No such file or directory
/opt/src/libreswan-3.15/programs/pluto/server.h:22:33: error: event2/event_struct.h: No such file or directory
In file included from /opt/src/libreswan-3.15/programs/pluto/demux.h:21,
                 from /opt/src/libreswan-3.15/programs/pluto/connections.c:60:
/opt/src/libreswan-3.15/programs/pluto/server.h:82: error: expected ‘)’ before ‘const’
/opt/src/libreswan-3.15/programs/pluto/server.h:83: error: expected ‘)’ before ‘ft’
In file included from /opt/src/libreswan-3.15/programs/pluto/connections.c:60:
/opt/src/libreswan-3.15/programs/pluto/demux.h:28: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘comm_handle_cb’
make[3]: *** [connections.o] Error 1
make[3]: Leaving directory `/opt/src/libreswan-3.15/OBJ.linux.x86_64/programs/pluto'
make[2]: *** [local-base] Error 2
make[2]: Leaving directory `/opt/src/libreswan-3.15/programs/pluto'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/opt/src/libreswan-3.15/programs'
make: *** [all] Error 2
```

**Comment by letoams** [comment]
You need libevent2-devel, see http://people.redhat.com/pwouters/libreswan_rhel6/

Rhe/centosl 6.8 and later will gain these packages soon

**Comment by hwdsl2** [comment]
Thank you for the quick reply. The provided link does not seem to work though. 

I compiled and installed the latest libevent2 from:
https://github.com/libevent/libevent/releases

After that Libreswan compiled and installed OK. However, when I tried to start ipsec, there is an error and it refuses to run:
```
Aug 26 17:36:12 my-host ipsec__plutorun: Starting Pluto subsystem...
Aug 26 17:36:12 my-host ipsec__plutorun: !pluto failure!:  exited with error status 127
Aug 26 17:36:12 my-host ipsec__plutorun: restarting IPsec after pause...
Aug 26 17:36:23 my-host ipsec__plutorun: Starting Pluto subsystem...
Aug 26 17:36:23 my-host ipsec__plutorun: !pluto failure!:  exited with error status 127
Aug 26 17:36:23 my-host ipsec__plutorun: restarting IPsec after pause...
```

**Comment by letoams** [comment]
http://people.redhat.com/pwouters/libreswan-rhel6/

**Comment by hwdsl2** [comment]
Can you please check this other issue regarding compile failure on Debian 7.0 (Wheezy):

```
root@server > make programs
... ...
cc  -m64 -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-all -fno-strict-aliasing -fPIE -DPIE -DFORCE_PR_ASSERT -DDNSSEC -DKLIPS -DLIBCURL -DUSE_MD5 -DHAVE_NM -DUSE_SHA2 -DUSE_SHA1 -DFIPSPRODUCTCHECK=\"/etc/system-fips\" -DIPSEC_CONF=\"/etc/ipsec.conf\" -DIPSEC_CONFDDIR=\"/etc/ipsec.d\" -DIPSEC_NSSDIR=\"/etc/ipsec.d\" -DIPSEC_CONFDIR=\"/etc\" -DIPSEC_EXECDIR=\"/usr/local/libexec/ipsec\" -DIPSEC_SBINDIR=\"/usr/local/sbin\" -DIPSEC_VARDIR=\"/var\" -DPOLICYGROUPSDIR=\"/etc/ipsec.d/policies\" -DSHARED_SECRETS_FILE=\"/etc/ipsec.secrets\" -DGCC_LINT -DALLOW_MICROSOFT_BAD_PROPOSAL -Werror -Wall -Wextra -Wformat -Wformat-nonliteral -Wformat-security -Wundef -Wmissing-declarations -Wredundant-decls -Wnested-externs -I/opt/src/libreswan-3.15/ports/linux/include -I/opt/src/libreswan-3.15/ports/linux/include -I/opt/src/libreswan-3.15/ports/linux/include -I/opt/src/libreswan-3.15/ports/linux/include  -I/opt/src/libreswan-3.15/programs/pluto/linux26 -I/opt/src/libreswan-3.15/include -I/opt/src/libreswan-3.15/lib/libcrypto -I/opt/src/libreswan-3.15/linux/include   -DUSE_KEYRR   -DNETKEY_SUPPORT -DKERNEL26_HAS_KAME_DUPLICATES -DPFKEY  -DUSE_TWOFISH -DUSE_SERPENT -DKLIPS -DPFKEY    -DUSE_AES -DUSE_3DES -DUSE_SHA2 -DUSE_SHA1 -DUSE_MD5 -DUSE_CAMELLIA   -DXAUTH_HAVE_PAM -DLIBCURL    -DHAVE_LIBCAP_NG -DHAVE_NM -I/usr/include/nss -I/usr/include/nspr     \
                -MMD -MF ./crypt_dbg.d \
                -o ./crypt_dbg.o \
                -c /opt/src/libreswan-3.15/programs/pluto/crypt_dbg.c
/opt/src/libreswan-3.15/programs/pluto/crypt_dbg.c: In function â€˜symkey_bytesâ€™:
/opt/src/libreswan-3.15/programs/pluto/crypt_dbg.c:102:2: error: implicit declaration of function â€˜PK11_Decryptâ€™ [-Werror=implicit-function-declaration]
/opt/src/libreswan-3.15/programs/pluto/crypt_dbg.c:102:2: error: nested extern declaration of â€˜PK11_Decryptâ€™ [-Werror=nested-externs]
cc1: all warnings being treated as errors
make[3]: *** [crypt_dbg.o] Error 1
make[3]: Leaving directory `/opt/src/libreswan-3.15/OBJ.linux.x86_64/programs/pluto'
make[2]: *** [local-base] Error 2
make[2]: Leaving directory `/opt/src/libreswan-3.15/programs/pluto'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/opt/src/libreswan-3.15/programs'
make: *** [all] Error 2
```

**Comment by letoams** [comment]
You need a newer nss version

**Comment by hwdsl2** [comment]
Thanks a lot for your help Paul. Today I see another compile error on CentOS 7 x86_64. Can you please look into this one as well?

```
cd . && flex -t ../../../lib/libipsecconf/$(basename ../../../lib/libipsecconf/parser.l) | grep -v '^extern int isatty' > lex.yy.c
cc -I/opt/src/libreswan-3.15/ports/linux/include -I/opt/src/libreswan-3.15/ports/linux/include -I/opt/src/libreswan-3.15/ports/linux/include -I/opt/src/libreswan-3.15/ports/linux/include -I. -I/opt/src/libreswan-3.15/linux/net/ipsec -I/opt/src/libreswan-3.15/linux/include -I/opt/src/libreswan-3.15 -I/opt/src/libreswan-3.15/include -I/usr/include/nss3 -I/usr/include/nspr4    -pthread  -m64 -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-all -fno-strict-aliasing -fPIE -DPIE -DFORCE_PR_ASSERT -DDNSSEC -DKLIPS -DLIBCURL -DUSE_MD5 -DHAVE_NM -DUSE_SHA2 -DUSE_SHA1 -DFIPSPRODUCTCHECK=\"/etc/system-fips\" -DIPSEC_CONF=\"/etc/ipsec.conf\" -DIPSEC_CONFDDIR=\"/etc/ipsec.d\" -DIPSEC_NSSDIR=\"/etc/ipsec.d\" -DIPSEC_CONFDIR=\"/etc\" -DIPSEC_EXECDIR=\"/usr/local/libexec/ipsec\" -DIPSEC_SBINDIR=\"/usr/local/sbin\" -DIPSEC_VARDIR=\"/var\" -DPOLICYGROUPSDIR=\"/etc/ipsec.d/policies\" -DSHARED_SECRETS_FILE=\"/etc/ipsec.secrets\" -DGCC_LINT -DALLOW_MICROSOFT_BAD_PROPOSAL -Werror -Wall -Wextra -Wformat -Wformat-nonliteral -Wformat-security -Wundef -Wmissing-declarations -Wredundant-decls -Wnested-externs \
        -MMD -MF ./lex.yy.d \
        -o ./lex.yy.o \
        -c /opt/src/libreswan-3.15/OBJ.linux.x86_64/lib/libipsecconf/lex.yy.c
<stdout>: In function â€˜yy_scan_bytesâ€™:
<stdout>:2169:17: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
cc1: all warnings being treated as errors
make[3]: *** [lex.yy.o] Error 1
make[3]: Leaving directory `/opt/src/libreswan-3.15/OBJ.linux.x86_64/lib/libipsecconf'
make[2]: *** [local-base] Error 2
make[2]: Leaving directory `/opt/src/libreswan-3.15/lib/libipsecconf'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/opt/src/libreswan-3.15/lib'
make: *** [all] Error 2
```

**Comment by letoams** [comment]
On Thu, 27 Aug 2015, Lin Song wrote:

> Thanks a lot for your help Paul. Today I see another compile error on CentOS 7 x86_64. Can you please look
> into this one as well?

>         -MMD -MF ./lex.yy.d \
>         -o ./lex.yy.o \
>         -c /opt/src/libreswan-3.15/OBJ.linux.x86_64/lib/libipsecconf/lex.yy.c
> <stdout>: In function â€˜yy_scan_bytesâ€™:
> <stdout>:2169:17: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
> cc1: all warnings being treated as errors

A workaround for now is to add:

WERROR_CFLAGS =""

to Makefile.inc (or Makefile.inc.local)

Paul

**Comment by hwdsl2** [comment]
Thank you! I added:
```
WERROR_CFLAGS =
```
to Makefile.inc.local, and it successfully compiled on CentOS 7.

**Comment by letoams** [comment]
fixed for next release to no longer need this workaround

**Comment by hwdsl2** [comment]
In Ubuntu 16.04 (to be released in a few days) and under GCC 5.3.1, compilation of Libreswan 3.17 fails with the following error:

<pre><code>        -o ./lex.yy.o \
        -c /opt/src/libreswan-3.17/OBJ.linux.x86_64/lib/libipsecconf/lex.yy.c
lex.yy.c.11447: In function 'yy_get_next_buffer':
lex.yy.c.11447:1723:44: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
cc1: all warnings being treated as errors
../../../mk/depend.mk:28: recipe for target 'lex.yy.o' failed
make[3]: *** [lex.yy.o] Error 1
make[3]: Leaving directory '/opt/src/libreswan-3.17/OBJ.linux.x86_64/lib/libipsecconf'
../../mk/library.mk:41: recipe for target 'local-base' failed
make[2]: *** [local-base] Error 2
make[2]: Leaving directory '/opt/src/libreswan-3.17/lib/libipsecconf'
../mk/subdirs.mk:33: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/opt/src/libreswan-3.17/lib'
/opt/src/libreswan-3.17/mk/subdirs.mk:33: recipe for target 'all' failed
make: *** [all] Error 2
</code></pre>

Thanks for looking into this issue again.

**Comment by letoams** [comment]
This seems to haunt us forever. It is tricky to keep supporting all combinations of flex/bison/gcc

You can set WERROR_FLAGS="" to work around it. (See mk/*)
